사측연산 계산기

0-jerry opened the pull request:
https://teamsparta.notion.site/Ch-1-1292dc3ef514805fad4be1b113588fa7

Int 타입의 사칙연산(`+`,`-`,`*`,`/`)을 지원하는 계산기 구현

0-jerry committed:
feat enum Operator

- 연산자 열거형

0-jerry committed:
Feat AddOperation

- 덧셈 연산을 위한 구조체
- calculate 메서드를 통해 연산값 반환

0-jerry committed:
Feat SubstractOperation

- 빼기 연산을 위한 구조체
- calculate 메서드를 통해 연산값 반환

0-jerry committed:
Fix 오탈자 수정

- AddOpreration -> AddOperation
- SubstractOpreration -> SubstractOperation

0-jerry committed:
Feat MultiplyOperation

- 곱하기 연산을 위한 구조체
- calculate 메서드를 통해 연산값 반환

0-jerry committed:
Feat DivideOperation

- 나누기 연산을 위한 구조체
- calculate 메서드를 통해 연산값 반환

0-jerry committed:
Feat calculator 메서드

- Calculator 구조체의 calculate 메서드를 통해 연산값 반환

0-jerry committed:
Refactor: AbstractOperation 프로토콜

- AbstractOperation 프로토콜
- calculator 의 파라미터 Int로 수정
- Reminder 나머지 연산 구조체

0-jerry committed:
Refactor: calculate 출력 형식 수정

- 연산의 형태 출력 추가
- 나눗셈 연산에 대한 나머지 값 출력 추가

0-jerry committed:
chore: 연산 테스트 코드 추가

0-jerry committed:
refactor: 출력 방식 수정

- AbstractOperation를 연산 구조체들의 calculate 메서드가 연산값을 출력
- Operator 열거형의 원시값 삭제
- ReminderOperation 나머지 연산 구조체 삭제
- DivideOperation 구조체의 calculate 메서드에서 secondNumber(나누는 수)가 0일 경우 "0으로는 나눌 수 없습니다." 출력
- 나눗셈 테스트 코드 추가

0-jerry committed:
Refator: calculate 정적메서드로 수정

0-jerry committed:
docs: add README

0-jerry committed:
docs:

0-jerry committed:
chore: Operator 열거형 주석

0-jerry committed:
Fix&Chore: 예외처리 추가, 출력형식 수정

- Int의 범위를 벗어나는 값에 대한 연산을 guard문을 통해 예외처리
- 출력 형식 수정

0-jerry committed:
Chore: 오탈자 수정, 정렬

0-jerry committed:
Refactor: MultiplyOperation.calculate

- guard문 리팩토링

0-jerry committed:
Refactor: 메서드 분리

O-O-wl commented on Calculator.playground/Contents.swift:
```diff
@@ -26,34 +26,45 @@ struct Calculator {
-        case substract
-        case multiply
-        case divide
+    private struct ReminderOperation: AbstractOperation {
```
아마 Remainer 일 것 같네요? 
IT에서는 Modular 연산이라고도 많이 명명하니 참고하세요 🙇‍♂️

O-O-wl commented on Calculator.playground/Contents.swift:
```diff
@@ -9,20 +9,26 @@ import Foundation
 struct Calculator {
     
-    func calculate(_ operator: Self.Operator, firstNumber: Double, secondNumber: Double ) -> Double {
+    func calculate(_ operator: Self.Operator, firstNumber: Int, secondNumber: Int ) {
```
기존 처럼 return 후, 프린트를 외부에서 찍는 게, 객체를 나누기도 좋고, Calculator를 유닛 테스트하기에 좋은 구조가 아닐까 하는 생각이 듭니답 🙇‍♂️

O-O-wl commented on Calculator.playground/Pages/Untitled Page.xcplaygroundpage/Contents.swift:
```diff
@@ -37,36 +37,97 @@ struct Calculator {
+            //합이 Int.max를 이하이며, Int.min이상
+            guard (firstNumber > 0 && Int.max - firstNumber >= secondNumber) ||
+                    (firstNumber < 0 && Int.min - firstNumber <= secondNumber) ||
+                    (firstNumber == 0) || (secondNumber == 0) else {
```
각 연산들을 조건을 나열하기보다, 이름으로 의미를 부여해보면 좋을 것 같네요
ex) n % 2 == 1 -> n.isOdd

peatle-e replied:
해당 조건문이 복잡하고 명료하지 않은 형태라 아쉬움이 있었는데
조건들을 더 명시적으로 표현할 수 있겠군요.
👍

O-O-wl commented on Calculator.playground/Pages/Untitled Page.xcplaygroundpage/Contents.swift:
```diff
@@ -37,36 +37,97 @@ struct Calculator {
+                } else {
+                    print("연산 범위를 초과하였습니다.")
+                }
+                return
```
이 부분을 함수로 분리해서 6줄의 코드를 설명하는 이름을 지어주면 좋겠네요

O-O-wl commented on Calculator.playground/Pages/Untitled Page.xcplaygroundpage/Contents.swift:
```diff
@@ -37,36 +37,97 @@ struct Calculator {
     
     // 곱셈 연산자 구조체
     private struct MultiplyOperation: AbstractOperation {
         static func calculate(_ firstNumber: Int,_ secondNumber: Int) {
```
계산 로직외의 피연산자들의 검증이 안에서 이뤄지는데요 
이 부분은 함수로 분리하는 건 어떨까요

peatle-e replied:
계산과 출력을 하나의 메서드에서 실행하면 검증이 어려워지는군요.
감사합니다.

O-O-wl reviewed (approved):
🙇‍♂️
고생하셨씁니다
마지막 커밋에서 특히나 개선이 많이 되었네요 

중간 중간에 수정한 걸 보니 고민의 흔적이 많이 보입니다
커밋도 나름 잘 나눈 것 같네요
아래의 코멘트를 한번 확인해보세요

0-jerry committed:
chore: 열거형, 오탈자, 메서드 수정

- Operator -> CalculatorOperator: 오인가능한 열거형 이름 수정
- substract -> subtract: 오탈자 수정
- checkUnberFlow -> checkUnderFlow: 오탈자 수정
- printResult -> formTo: 출력을 분리하고, String값을 반환
